Tribler/Core/TorrentChecker/session.py

```diff
@@ -198,309 +185,384 @@ def __init__(self, tracker_url, tracker_address, announce_page, on_result_callba
+        self.cleanup()
 
-        # the remaining part
-        else:
-            self._message_buffer += response
-            self._received_length += len(response)
-
-        # check the read count
-        if self._received_length >= self._content_length:
-            # process the retrieved information
-            success = self._process_scrape_response()
-            if success:
-                self._is_finished = True
-            else:
-                self._is_failed = True
-
-        # wait for more
-        else:
-            pass
-
-    def _process_header(self):
-        # get and check HTTP response code
-        protocol, code, msg = self._header_buffer.split(' ', 2)
-        if code == '301' or code == '302':
-            idx = self._header_buffer.find('Location: ')
-            if idx == -1:
-                self._is_failed = True
-            else:
-                new_location = (self._header_buffer[idx:].split('\r\n')[0]).split(' ')[1]
-                try:
-                    idx = new_location.find('info_hash=')
-                    if idx != -1:
-                        new_location = new_location[:idx]
-                    if new_location[-1] != '/':
-                        new_location += "/"
-                    new_location += "announce"
-
-                    tracker_type, tracker_address, announce_page = parse_tracker_url(new_location)
-                    if tracker_type != self._tracker_type:
-                        raise RuntimeError(u"cannot redirect to a different tracker type: %s", new_location)
-
-                    else:
-                        self._logger.debug(u"%s being redirected to %s", self, new_location)
-
-                        self._tracker_address = tracker_address
-                        self._announce_page = announce_page
-                        self._socket.close()
-                        self._socket = None
-
-                        self.recreate_connection()
-
-                except RuntimeError as run_err:
-                    self._logger.info(u"%s: Runtime Error: %s, address: %s, announce: %s",
-                                      self, run_err, self._tracker_address, self._announce_page)
-                    self._is_failed = True
-
-                except Exception as err:
-                    self._logger.exception(u"Failed to process HTTP tracker header: [%s], Tracker: %s,"
-                                           u" Tracker Address: %s, Tracker Announce: %s",
-                                           err, self._tracker_url, self._tracker_address, self._announce_page)
-                    self._logger.debug(u"Header: %s", self._header_buffer)
-                    self._is_failed = True
+    def _process_scrape_response(self, body):
+        """
+        This function handles the response body of a HTTP tracker,
+        parsing the results.
+        """
+        # parse the retrieved results
+        if body is None:
+            self.failed()
```

Comment:
@whirm would be raising an exception and catching that with the errback of the chain be a good solution to this problem? the `self.failed` can then be added as an errback to that parseBody deferred.